pkg/provision/gce.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProvisionedHost{
+		ID: constructCustomID(host),
+	}, nil
+}
+
+// Delete removes the GCE instance when the LoadBalancer is deleted
+func (p *GCEProvisioner) Delete(id string) error {
+	projectID, zone, hostname := getFieldsFromID(id)
+	_, err := p.computeService.Instances.Delete(projectID, zone, hostname).Do()
+	return err
+}
+
+// Status checks the status of the provisioned GCE instance
+func (p *GCEProvisioner) Status(id string) (*ProvisionedHost, error) {
+	projectID, zone, hostname := getFieldsFromID(id)
+	instance, err := p.computeService.Instances.Get(projectID, zone, hostname).Do()
+	if err != nil {
+		return nil, fmt.Errorf("could not get GCE instance: %v", err)
+	}
+
+	ip := ""
+
+	for _, networkInterface := range instance.NetworkInterfaces {
+		for _, accessConfig := range networkInterface.AccessConfigs {
+			if accessConfig.NatIP != "" {
+				ip = accessConfig.NatIP
+			}
+		}
+	}
+
+	status := instance.Status
+	if instance.Status == "RUNNING" {
+		status = "active"
+	}
+
+	return &ProvisionedHost{
+		IP:     ip,
+		ID:     id,
+		Status: status,
+	}, nil
+}
+
+// CheckFirewallRuleExists checks if the 'firewallRuleName' exist in the firewall rule exists
+func (p *GCEProvisioner) CheckFirewallRuleExists(projectID string, firewallRuleName string, firewallPort string) (bool, error) {
+	firewallList, err := p.computeService.Firewalls.List(projectID).Do()
+	if err != nil {
+		return false, err
+	}
+	for _, firewallFromList := range firewallList.Items {
+		if firewallFromList.Name == firewallRuleName {
+			for _, allowed := range firewallFromList.Allowed {
+				for _, port := range allowed.Ports {
+					if port == firewallPort {
+						return true, nil
+					}
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// CreateInletsFirewallRule allows to create a firewallRule to allow for the inlets control port
+func (p *GCEProvisioner) CreateInletsFirewallRule(projectID, firewallRuleName, firewallPort string) (*compute.Operation, error) {
+	firewall := &compute.Firewall{
+		Name:    firewallRuleName,
+		Network: fmt.Sprintf("projects/%s/global/networks/default", projectID),
+		Allowed: []*compute.FirewallAllowed{
+			{
+				IPProtocol: "tcp",
+				Ports:      []string{firewallPort},
+			},
+		},
+		SourceRanges: []string{"0.0.0.0/0"},
+		Direction:    "INGRESS",
+		TargetTags:   []string{"inlets"},
+	}
+	firewallInsetOp, err := p.computeService.Firewalls.Insert(projectID, firewall).Do()
+	return firewallInsetOp, err
+}
+
+// Just a hack to get a custom ID that we can use to get the details like projectID, zone, and instance name
+func constructCustomID(host BasicHost) string {
+	return fmt.Sprintf("%s|%s|%s", host.Additional["project_id"], host.Additional["zone"], host.Name)
+}
+
+// return the vals from ID construct with constructCustomID
+func getFieldsFromID(id string) (projectID, zone, hostname string) {
+	fields := strings.Split(id, "|")
```

Comment:
Add range checking please, and return an error along with the returned value

Reply:
Sure!